File: 15May-2026/Day1/FindDuplicates.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Random;
import java.util.Set;

public class FindDuplicates {

    /**
     * Generate a list of random numbers (with possible duplicates).
     *
     * @param size  how many numbers to generate
     * @param start minimum value (inclusive)
     * @param end   maximum value (inclusive)
     * @return list of random integers
     */
    public static List<Integer> generateRandomNumbers(int size, int start, int end) {
        Random random = new Random();
        List<Integer> numbers = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            numbers.add(random.nextInt(end - start + 1) + start);
        }
        return numbers;
    }

    /**
     * Find duplicate numbers in a list.
     *
     * @param numbers input list
     * @return sorted list of duplicates
     */
    public static List<Integer> findDuplicates(List<Integer> numbers) {
        Set<Integer> seen = new HashSet<>();
        Set<Integer> duplicates = new HashSet<>();

        for (int num : numbers) {
            if (!seen.add(num)) {
                duplicates.add(num);
            }
        }

        List<Integer> result = new ArrayList<>(duplicates);
        Collections.sort(result);
        return result;
    }

    public static void main(String[] args) {
        List<Integer> numbers = generateRandomNumbers(20, 1, 15);
        System.out.println("Generated numbers : " + numbers);

        List<Integer> duplicates = findDuplicates(numbers);

        if (!duplicates.isEmpty()) {
            System.out.println("Duplicate numbers : " + duplicates);
        } else {
            System.out.println("No duplicates found.");
        }
    }
}

File: 15May-2026/Day1/find_duplicates.rb
require 'set'

# Generate an array of random numbers (with possible duplicates).
# @param size  [Integer] how many numbers to generate
# @param start [Integer] minimum value (inclusive)
# @param stop  [Integer] maximum value (inclusive)
# @return [Array<Integer>]
def generate_random_numbers(size: 20, start: 1, stop: 15)
  Array.new(size) { rand(start..stop) }
end

# Find duplicate numbers in an array.
# @param numbers [Array<Integer>]
# @return [Array<Integer>] sorted list of duplicates
def find_duplicates(numbers)
  seen       = Set.new
  duplicates = Set.new

  numbers.each do |num|
    if seen.include?(num)
      duplicates.add(num)
    else
      seen.add(num)
    end
  end

  duplicates.to_a.sort
end

# --- main ---
numbers = generate_random_numbers
puts "Generated numbers : #{numbers}"

duplicates = find_duplicates(numbers)

if duplicates.any?
  puts "Duplicate numbers : #{duplicates}"
else
  puts "No duplicates found."
end
